### ISSUE TITLE
Minimize dependencies in virtual environment

### ISSUE THREAD

**Issue by nietootein** [bryankim96/ctalearn#13]
We should filter out from requirements.txt all those dependencies that are not actually necessary for ctalearn to properly run, so the installation is lighter and faster. Some users have requested instructions for a clean uninstall that will remove not just the virtual environment but also the packages that were downloaded to set it up (without conflicting with other environments).

**Comment by aribrill** [comment]
The only actual direct requirements are those listed under "Dependencies" in the Readme. In addition, there is an optional requirement, PIL, which is used only in the script visualize_bounding_boxes.py and is not needed for standard functionality.

**Comment by nietootein** [comment]
Hi all,

It turns out that pip does not feature a dependency resolver. If we want to keep supporting the installation of CTALearn via pip (i.e. without using conda) we have to keep a long list of dependencies, and likely pinned ones. I've been figuring out the minimal set of dependencies in our dependencies.txt if CTALearn is installed via the recommended way (w/ Anaconda, into a virtual environment) and it boils down to: numpy, pytables, opencv, python=3.6 and tensorflow-gpu=1.8. Such a simple file of dependencies provides us with a working virtual environment for both linux-64 (tested under Ubuntu 16.04) and osx-64 (tested under OSX 10.13), thanks to the dependency resolver in conda. From the perspective of code maintenance it seems easier to me to keep a minimal collection of dependencies in our dependencies.txt and let conda to resolve the remaining dependencies for us, rather than maintaining a long list of dependencies. In addition, the removal instructions are fairly trivial and safe if the installation was performed via conda... I know requesting people to install Anaconda in order to run CTALearn may be a nuisance to some potential user, but it may help to ease the package maintenance... What's your take on this? I guess in the long run we should think about packaging the code and upload it to PyPI?

Cheers,

D.

**Comment by nietootein** [comment]
Related to #31.

**Comment by aribrill** [comment]
I think installation via Anaconda with only the direct required packages listed in the requirements file is a good approach. It should certainly make uninstallation easier. We could add to the list of dependencies PIL, which is an optional dependency only for visualize_bounding_boxes.py as mentioned above. I think that since ctapipe is installed through Anaconda (at least for the user/non-development version), that is a pretty good precedent for requesting users to do this. Packaging and uploading the code in future versions makes sense to me.

**Comment by nietootein** [comment]
Done. I've tested the installation (both CPU and GPU flavors) on a linux-x64 system and on a osx-64 system (CPU flavor only, for a lack of GPU on that system). I also updated the installation instructions accordingly (and removed the section dealing with the installation w/ pip) as well as added a few words on how to remove the package and dependencies from your system. I guess this issue and #31 can be closed now.
